### PR TITLE
Bluetooth: audio: ascs: Fix repeated ops->released callback call

### PR DESCRIPTION
**Pull request by MariuszSkamra**
Once the ACL disconnects, the implementation releases the related
endpoints. The ops->released callback is called once ASE enters IDLE
state, thus there is no need to call it explicitly.
This fixes repeated ops->released callback call.

### CODE REVIEW

File: subsys/bluetooth/audio/ascs.c [hunk 1]
```diff
@@ -954,23 +953,6 @@ static void disconnected(struct bt_conn *conn, uint8_t reason)
...
-				ops = stream->ops;
-				if (ops != NULL && ops->released != NULL) {
-					ops->released(stream);
-				} else {
-					LOG_WRN("No callback for released set");
-				}
-			}
-		}
-
-		if (stream != NULL && stream->conn != NULL) {
```
Comment (fredrikdanebjer) [inline subsys/bluetooth/audio/ascs.c:971]: Why do you remove the unref here?

Comment (MariuszSkamra) [reply subsys/bluetooth/audio/ascs.c:971]: Because `bt_bap_stream_reset` called from `ascs_ep_set_state(BT_BAP_EP_STATE_IDLE)` does the same job.

File: subsys/bluetooth/audio/ascs.c [hunk 2]
```diff
@@ -954,23 +953,6 @@ static void disconnected(struct bt_conn *conn, uint8_t reason)
...
-				const struct bt_bap_stream_ops *ops;
-
-				/* Notify upper layer */
-				ops = stream->ops;
-				if (ops != NULL && ops->released != NULL) {
-					ops->released(stream);
-				} else {
-					LOG_WRN("No callback for released set");
-				}
-			}
```
Comment (Thalley) [inline subsys/bluetooth/audio/ascs.c:968]: I might be missing something, but removing this may not be correct. 

As far as I can tell, the flow is

1. ACL Disconnects -> `disconnected`
2. `disconnected` -> `ase_release` if not `IDLE`
3. `ase_release` -> `ascs_ep_set_state(&ase->ep, BT_BAP_EP_STATE_RELEASING);`
4. `ascs_ep_set_state(&ase->ep, BT_BAP_EP_STATE_RELEASING);` -> `ascs_ep_set_state(ep, BT_BAP_EP_STATE_IDLE);` if `BT_ISO_STATE_DISCONNECTED` else `ascs_disconnect_stream`

In other words, `ops->released` will only be called here if the CIS disconnects before the ACL (which is not a given). 

Your assumption (https://github.com/zephyrproject-rtos/zephyr/pull/57835/files#r1193347864) that `ase_release` will always call `ascs_ep_set_state(BT_BAP_EP_STATE_IDLE)` is incorrect

Comment (MariuszSkamra) [reply subsys/bluetooth/audio/ascs.c:968]: You're right, my comment (https://github.com/zephyrproject-rtos/zephyr/pull/57835/files#r1193347864) is a bit simplified, however the `ascs_ep_set_state(BT_BAP_EP_STATE_IDLE)` will be called eventually anyway.
Let me complete your flow:

> 1. ACL Disconnects -> `disconnected`
> 2. `disconnected` -> `ase_release` if not `IDLE`
> 3. `ase_release` -> `ascs_ep_set_state(&ase->ep, BT_BAP_EP_STATE_RELEASING);`
4a. if `BT_ISO_STATE_DISCONNECTED`: `ascs_ep_set_state(ep, BT_BAP_EP_STATE_IDLE);`
4b. else `ascs_disconnect_stream`
5b. `ascs_iso_disconnected` -> `ascs_ep_set_state(ep, BT_BAP_EP_STATE_IDLE);` if BT_BAP_EP_STATE_RELEASING

Comment (MariuszSkamra) [reply subsys/bluetooth/audio/ascs.c:1012]: @Thalley I've added unit test for this particular case `test_release_ase_on_acl_disconnection_server_terminates_cis`

Comment (Thalley) [reply subsys/bluetooth/audio/ascs.c:1012]: > ascs_ep_set_state(BT_BAP_EP_STATE_IDLE) will be called eventually anyway.

Agreed, but the difference is whether or not the stream object is reusable right after the application gets the ACL disconnect callback. 

The code to release the stream here was added for that purpose, and this would have broken then shell earlier as it expected the streams to be released at ACL disconnect. 

Now that I think about it, since we do not know the order of the app vs ASCS ACL disconnect callback, the application can still not expect the streams to be released by the time the callback gets called.

I am OK with this change, and I see that the shell no longer triggers the unicast group delete on ACL disconnect, but rather on stream released.

File: tests/bluetooth/audio/ascs/src/main.c [hunk 3]
```diff
@@ -220,3 +220,79 @@ ZTEST_F(ascs_test_suite, test_abort_client_operation_if_callback_not_registered)
...
+	}
+
+	zexpect_not_null(ase);
+	zexpect_true(ase_id != 0x00);
+
+	bt_bap_unicast_server_register_cb(&mock_bap_unicast_server_cb);
+
+	/* Set ASE to non-idle state */
+	test_preamble_state_streaming(conn, ase_id, stream, &chan,
+				      !IS_ENABLED(CONFIG_BT_ASCS_ASE_SNK));
```
Comment (Thalley) [inline tests/bluetooth/audio/ascs/src/main.c:247]: Instead of relying on `IS_ENABLED(CONFIG_BT_ASCS_ASE_SNK)` would it make more sense to test this with
1) Sink stream in CIS only
2) Source stream in CIS only
3) Sink and Source stream in same CIS
?

File: tests/bluetooth/audio/ascs/src/main.c [hunk 4]
```diff
@@ -220,3 +220,79 @@ ZTEST_F(ascs_test_suite, test_abort_client_operation_if_callback_not_registered)
...
+
+	/* Mock ACL disconnection */
+	mock_bt_conn_disconnected(conn);
+
+	/* Server waits for CIS to be disconnected by client */
+	k_sleep(K_MSEC(CONFIG_BT_ASCS_ISO_DISCONNECT_DELAY));
+
+	/* Expected to notify the upper layers */
+	expect_bt_bap_unicast_server_cb_release_called_once(stream);
+	expect_bt_bap_stream_ops_released_called_once(stream);
```
Comment (Thalley) [inline tests/bluetooth/audio/ascs/src/main.c:295]: The `CONFIG_BT_ASCS_ISO_DISCONNECT_DELAY` is only used to disconnect the CIS if the stream goes out of the streaming state; it is not related to CIS disconnect events due to ACL disconnects. 

I don't see the value of this specific test case, as neither the client nor the server initiates the CIS disconnect if the ACL disconnects

File: tests/bluetooth/audio/mocks/src/conn.c [hunk 5]
```diff
@@ -29,3 +29,12 @@ void bt_conn_unref(struct bt_conn *conn)
...
 }
+
+void mock_bt_conn_disconnected(struct bt_conn *conn)
+{
+	STRUCT_SECTION_FOREACH(bt_conn_cb, cb) {
+		if (cb->disconnected) {
+			cb->disconnected(conn, 0x13);
+		}
+	}
+}
```
Comment (Thalley) [inline tests/bluetooth/audio/mocks/src/conn.c:40]: Consider add the reason as a parameter here; we may want to test with both `BT_HCI_ERR_REMOTE_USER_TERM_CONN` and `BT_HCI_ERR_CONN_TIMEOUT`

File: tests/bluetooth/audio/mocks/src/kernel.c [hunk 6]
```diff
@@ -54,6 +56,28 @@ int k_work_cancel_delayable(struct k_work_delayable *dwork)
...
 
+int32_t k_sleep(k_timeout_t timeout)
+{
+	struct k_work *work;
+
+	SYS_SLIST_FOR_EACH_CONTAINER(&work_pending, work, node) {
+		if (work->flags & K_WORK_DELAYED) {
+			struct k_work_delayable *dwork;
+
+			dwork = CONTAINER_OF(work, struct k_work_delayable, work);
```
Comment (Thalley) [inline tests/bluetooth/audio/mocks/src/kernel.c:67]: Not a change needed, but could consider using `k_work_delayable_from_work`

**Review by Thalley — CHANGES_REQUESTED** [review]

File: tests/bluetooth/audio/ascs/src/main.c (hunk 4 above)
Comment (MariuszSkamra) [reply tests/bluetooth/audio/ascs/src/main.c:295]: Cannot agree, as ACL disconnection triggers the `ascs_disconnect_stream` if CIS has been not disconnected before ACL (which is possible case - allowed by the spec).

**Comment by MariuszSkamra** [comment]
Added 2 more test cases `test_release_stream_pair_on_acl_disconnection_client_terminates_cis` and `test_release_stream_pair_on_acl_disconnection_server_terminates_cis` to verify whether stream pair is released on ACL disconnection

File: tests/bluetooth/audio/ascs/src/main.c (hunk 4 above)
Comment (Thalley) [reply tests/bluetooth/audio/ascs/src/main.c]: Hmm, I see your point, however there is not really any value in attempting to disconnect the CIS _after_ the ACL disconnect event: We know the CIS will get disconnected right after (if not before, depending on the controller implementation). 

IMO the host should not attempt to disconnect the CIS on ACL disconnect, which I can see that we do in ASCS - Maybe that part needs changed? There no value in calling `ascs_disconnect_stream` if ACL state is not `BT_CONN_STATE_CONNECTED`.

File: tests/bluetooth/audio/ascs/src/main.c [hunk 7]
```diff
@@ -220,3 +221,203 @@ ZTEST_F(ascs_test_suite, test_abort_client_operation_if_callback_not_registered)
...
+	test_ase_control_client_config_qos(conn, ase_src_id);
+	test_ase_control_client_enable(conn, ase_src_id);
+
+	err = mock_bt_iso_accept(conn, 0x01, 0x01, &chan);
+	zassert_equal(0, err, "Failed to connect iso: err %d", err);
+
+	test_ase_control_client_receiver_start_ready(conn, ase_src_id);
+
+	err = bt_bap_stream_start(&snk_stream);
+	zassert_equal(0, err, "bt_bap_stream_start err %d", err);
```
Comment (Thalley) [inline tests/bluetooth/audio/ascs/src/main.c:344]: Is it not possible to use `test_preamble_state_streaming` here?

File: tests/bluetooth/audio/mocks/include/bap_unicast_server_expects.h [hunk 8]
```diff
@@ -155,6 +155,26 @@ do {
 				       func_name, "stream");))                                     \
 } while (0)
 
+#define expect_bt_bap_unicast_server_cb_release_called_twice(_streams)                             \
```
Comment (Thalley) [inline tests/bluetooth/audio/mocks/include/bap_unicast_server_expects.h:158]: This seems like it'll scale poorly :D 

Does it make sense to have a `expect_bt_bap_unicast_server_cb_release_called(_cnt, _streams)`? as a more generic approach? 
Not just for this, for for all the `expect_..._called_...`?

File: tests/bluetooth/audio/mocks/include/bap_unicast_server_expects.h [hunk 9]
```diff
@@ -155,6 +155,26 @@ do {
...
+do {                                                                                               \
+	const char *func_name = "bt_bap_unicast_server_cb.release";                                \
+												   \
+	zassert_equal(2, mock_bap_unicast_server_cb_release_fake.call_count,                       \
+		      "'%s()' was called %u times, but expected once",                             \
```
Comment (Thalley) [inline tests/bluetooth/audio/mocks/include/bap_unicast_server_expects.h:163]:
```suggestion
		      "'%s()' was called %u times, but expected twice",                             \
```

File: tests/bluetooth/audio/ascs/src/main.c (hunk 7 above)
Comment (MariuszSkamra) [reply tests/bluetooth/audio/ascs/src/main.c:344]: This requires refactor as well, I'll create an issue instead

**Review by Thalley — DISMISSED** [review]

File: tests/bluetooth/audio/ascs/src/main.c [hunk 10]
```diff
@@ -220,3 +221,203 @@ ZTEST_F(ascs_test_suite, test_abort_client_operation_if_callback_not_registered)
...
 	zassert_equal(0x00, param->reason, "unexpected Reason 0x%02x", param->reason);
 }
+
+ZTEST_F(ascs_test_suite, test_release_ase_on_acl_disconnection_client_terminates_cis)
+{
+	const struct bt_gatt_attr *ase;
+	struct bt_bap_stream *stream = &fixture->stream;
+	struct bt_conn *conn = &fixture->conn;
+	struct bt_iso_chan *chan;
+	uint8_t ase_id;
```
Comment (szymon-czapracki) [inline tests/bluetooth/audio/ascs/src/main.c:231]: Keep definitions from longest to shortest for code style conformity.
Applies in multiple places.

**Review by szymon-czapracki — CHANGES_REQUESTED** [review]

File: tests/bluetooth/audio/ascs/src/main.c (hunk 10 above)
Comment (MariuszSkamra) [reply tests/bluetooth/audio/ascs/src/main.c]: done

**Review by szymon-czapracki — APPROVED** [review]

**Review by sjanc — APPROVED** [review]

**Review by Thalley — APPROVED** [review]
